### ISSUE TITLE
systemd-networkd Address=0.0.0.0/30 assigns too many addresses

### ISSUE THREAD

**Issue by webstrand** [systemd/systemd#26113]
### systemd version the issue has been seen with

252

### Used distribution

OpenSuse Tumbleweed

### Linux kernel version used

6.1.4-1-default

### CPU architectures issue was seen on

x86_64

### Component

networkctl

### Expected behaviour you didn't see

Once the `Address=0.0.0.0/30` has assigned an IP address, it does not assign another one.

### Unexpected behaviour you saw

At configure time `Address=0.0.0.0/30` configures two addresses. Over time more address ranges are added. Previous assignments are not recognized.

```
4: ve-test@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group 666 qlen 1000
    link/ether 3a:45:7a:bd:32:68 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.82.65/30 brd 192.168.82.67 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.43.133/30 brd 192.168.43.135 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.113.117/30 brd 192.168.113.119 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.159.21/30 brd 192.168.159.23 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.254.45/30 brd 192.168.254.47 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.69.141/30 brd 192.168.69.143 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.90.225/30 brd 192.168.90.227 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.187.197/30 brd 192.168.187.199 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.155.181/30 brd 192.168.155.183 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.138.217/30 brd 192.168.138.219 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.215.193/30 brd 192.168.215.195 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.145.253/30 brd 192.168.145.255 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.109.105/30 brd 192.168.109.107 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.31.17/30 brd 192.168.31.19 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.33.85/30 brd 192.168.33.87 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.159.5/30 brd 192.168.159.7 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.66.13/30 brd 192.168.66.15 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.161.49/30 brd 192.168.161.51 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.32.65/30 brd 192.168.32.67 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.13.117/30 brd 192.168.13.119 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.133.85/30 brd 192.168.133.87 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.222.153/30 brd 192.168.222.155 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.174.53/30 brd 192.168.174.55 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.97.241/30 brd 192.168.97.243 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.3.69/30 brd 192.168.3.71 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.84.197/30 brd 192.168.84.199 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.232.105/30 brd 192.168.232.107 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.229.113/30 brd 192.168.229.115 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.211.229/30 brd 192.168.211.231 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.20.61/30 brd 192.168.20.63 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.157.225/30 brd 192.168.157.227 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.34.69/30 brd 192.168.34.71 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.128.129/30 brd 192.168.128.131 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.2.213/30 brd 192.168.2.215 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.247.165/30 brd 192.168.247.167 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.2.113/30 brd 192.168.2.115 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.121.61/30 brd 192.168.121.63 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.5.209/30 brd 192.168.5.211 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.239.113/30 brd 192.168.239.115 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.254.65/30 brd 192.168.254.67 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.154.121/30 brd 192.168.154.123 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.143.81/30 brd 192.168.143.83 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.87.21/30 brd 192.168.87.23 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.2.153/30 brd 192.168.2.155 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.102.109/30 brd 192.168.102.111 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.209.57/30 brd 192.168.209.59 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.210.245/30 brd 192.168.210.247 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.111.169/30 brd 192.168.111.171 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.155.37/30 brd 192.168.155.39 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.71.49/30 brd 192.168.71.51 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.64.221/30 brd 192.168.64.223 scope global ve-test
       valid_lft forever preferred_lft forever
       
```

### Steps to reproduce the problem

1. `machinectl pull-tar --verify=no http://download.opensuse.org/tumbleweed/appliances/opensuse-tumbleweed-image.x86_64-networkd.tar.xz tumbleweed`
2. `machinectl clone tumbleweed test`
3. `machinectl start test`
4. Observe that the `ve-test` interface was created and has been configured by `systemd-networkd`.
5. Observe that it has _at least_ two configured IP addresses.

```
9: ve-test@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group 666 qlen 1000
    link/ether d6:4b:72:ab:56:41 brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet 192.168.187.137/30 brd 192.168.187.139 scope global ve-test
       valid_lft forever preferred_lft forever
    inet 192.168.150.225/30 brd 192.168.150.227 scope global ve-test
       valid_lft forever preferred_lft forever
```

### Additional program output to the terminal or log subsystem illustrating the issue

This is the `/etc/systemd/network/79-container-ve.network` that is configuring my `ve-` devices. The builtin `80-container-ve.network` has the same problem.

```toml
[Match]
Name=ve-*
Driver=veth

[Network]
Address=0.0.0.0/30
LinkLocalAddressing=no
LLDP=no
LLMNR=no
IPForward=yes
DHCPServer=yes
IPv6AcceptRA=no
IPv6SendRA=no
```

**Comment by webstrand** [comment]
Here's the `systemd-networkd` debug log generated when `ve-test` is created:

https://gist.github.com/webstrand/e79b3187f1351290d306b1f7533d7b98

the logs seemed to indicate that `address_pool_acquire_one` gets called twice, once when networkd initially configures the interface and once again when it gains carrier (when the other side of the veth is brought UP, I think?).

**Comment by yuwata** [comment]
Ouch. Indeed...

**Comment by bluca** [comment]
@yuwata any update on this?

**Comment by yuwata** [comment]
Not yet. Moved to the next milestone.

**Comment by yuwata** [comment]
Fix is waiting in #28287. Moving this back to v254 milestone.
